src/pcp.rs

```diff
@@ -135,65 +199,269 @@ pub trait ValueParam: Clone + Debug {
+            for wire in 0..gadget.arity() {
+                discrete_fourier_transform(&mut f[wire], &gadget.f_vals[wire], m)?;
+                discrete_fourier_transform_inv_finish(&mut f[wire], m, m_inv);
+
+                // The first point on each wire polynomial is a random value chosen by the prover. This
+                // point is stored in the proof so that the verifier can reconstruct the wire
+                // polynomials.
+                proof[proof_len + wire] = gadget.f_vals[wire][0];
+            }
+
+            // Construct the gadget polynomial `G(f[0], ..., f[g_arity-1])` and append it to `proof`.
+            gadget.call_poly(&mut proof[proof_len + gadget.arity()..], &f)?;
+            proof_len += gadget.arity() + gadget.degree() * (m - 1) + 1;
+        }
 
-    /// Evaluates the validity circuit on the given input (i.e., `self`) and returns the output.
-    /// `joint_rand` is the joint randomness shared by the prover and verifier. `g` is the sequence
-    /// of gadgets called by the circuit.
+        // Truncate the buffer to the size of the proof.
+        proof.truncate(proof_len);
+        Ok(proof)
+    }
+
+    /// Query an input and proof and return the verifier message.
     ///
-    /// ```
-    /// use prio::pcp::types::Count;
-    /// use prio::pcp::{Value, ValueParam};
-    /// use prio::field::{random_vector, FieldElement, Field64};
+    /// # Parameters
     ///
-    /// let x: Count<Field64> = Count::new(1).unwrap();
-    /// let joint_rand = random_vector(x.param().joint_rand_len()).unwrap();
-    /// let v = x.valid(&mut x.gadget(), &joint_rand).unwrap();
-    /// assert_eq!(v, Field64::zero());
-    /// ```
-    fn valid(
+    /// * `input` is the input or input share.
+    /// * `proof` is the proof or proof share.
+    /// * `query_rand` is the verifier's randomness.
+    /// * `joint_rand` is the randomness shared by the prover and verifier.
+    /// * `num_shares` is the total number of input shares. Typically the value of this parameter
+    /// is `1`, but [`prio3`](crate::vdaf::prio3) sets this to the number of aggregators.
+    fn query(
         &self,
-        gadgets: &mut Vec<Box<dyn Gadget<Self::Field>>>,
+        input: &[Self::Field],
+        proof: &[Self::Field],
+        query_rand: &[Self::Field],
         joint_rand: &[Self::Field],
-    ) -> Result<Self::Field, PcpError>;
+        num_shares: usize,
+    ) -> Result<Vec<Self::Field>, PcpError> {
+        let mut proof_len = 0;
+        let mut shim = self
+            .gadget()
+            .into_iter()
+            .enumerate()
+            .map(|(idx, gadget)| {
+                if idx >= query_rand.len() {
+                    return Err(PcpError::Query(format!(
+                        "short query randomness: got {}; want {}",
+                        query_rand.len(),
+                        self.query_rand_len()
+                    )));
+                }
```

Comment:
Would it be simpler to do `if self.gadget.len() == query_rand.len() {}` once, before iterating `self.gadget()`? Then I think you could also drop the check for `query_rand.len() > shim.len()` on line 364, since `shim.len()` should always be the same as `self.gadget.len()`.

 Additionally I think you could zip `self.gadget().into_iter()` with `query_rand.into_iter()` and avoid indexing into `query_rand`, and then you don't need to `.enumerate()`, but either way is fine with me.